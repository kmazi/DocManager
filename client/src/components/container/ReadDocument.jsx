import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import { changeTitleValue, editDocument } from '../../actions/documentActions';
import ReadDocument from '../presentation/ReadDocument';

const mapStateToProps = state => ({
  documentTitle: state.readDocument.documentTitle,
  documentId: state.readDocument.document.id,
  userId: state.authenticateUser.userId,
  ownerId: state.readDocument.document.userId,
  roleType: state.authenticateUser.roleType,
  docStatus: state.readDocument.docStatus,
  body: state.readDocument.document.body,
  author: state.readDocument.document.author,
  modifiedDate: state.readDocument.document.updatedAt,
});
export default connect(mapStateToProps,
  { changeTitleValue, editDocument })(withRouter(ReadDocument));
