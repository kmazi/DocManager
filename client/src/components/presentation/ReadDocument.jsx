import React from 'react';
import propTypes from 'prop-types';
import parser from 'html-react-parser';
import $ from 'jquery';
import Alert from 'sweetalert2';
import TinyMCE from 'react-tinymce';

let documentBody = '';
const documentValue = {};
const formatDate = (datetime) => {
  if (datetime) {
    const date = datetime.substring(0, datetime.indexOf('T'));
    const formattedDate = new Date(date);
    return formattedDate.toDateString();
  }
};

const processEdit = (editDocument, documentId, history) => {
  const title = $('#title').val();
  if (title !== '') {
    documentValue.title = title;
  }
  if (documentBody !== '') {
    documentValue.body = documentBody;
  }
  editDocument(documentValue, documentId)
    .then((res) => {
      if (res.status === 'successful') {
        Alert({
          title: 'Update successful',
          text: 'You have successfully updated your document',
          type: 'success',
          confirmButtonText: 'ok'
        });
        history.push('/user/documents');
      } else {
        Alert({
          title: 'Error updating document',
          text: res.message,
          type: 'error',
          confirmButtonText: 'ok'
        });
      }
    });
};

const handleEditorChange = (event) => {
  documentBody = event.target.getContent();
};

const ReadDocument = ({ documentTitle, body, author,
  modifiedDate, history, changeTitleValue, editDocument,
  documentId, docStatus, userId, roleType, ownerId }) => (
    <div id="readdocument" className="container">
      <input
        className="center-align"
        id="title"
        type="text"
        onChange={() => {
          changeTitleValue($('#title').val());
        }}
        value={documentTitle}
        disabled
      />
      <div className="" id="docbody">
        {parser(body)}
      </div><br />
      <div id="docbodyedit" className="hide">
        <TinyMCE
          content={body}
          config={{
            plugins: 'autolink link image lists print preview',
            toolbar: 'undo redo | bold italic | alignleft aligncenter alignright',
            height: 280,
            width: '100%',
            browser_spellcheck: true,
          }}
          onChange={handleEditorChange}
        /><br />
      </div>
      <div>
        <span>Author: {author}</span><br />
        <span>Modified: {formatDate(modifiedDate)}</span>
        <button
          className="btn right"
          onClick={(event) => {
            event.preventDefault();
            history.push('/user/documents');
          }}
        >back</button>
        <button
          id="editbtn"
          style={{ display: userId === ownerId || roleType === 'Admin'
          || roleType === 'SuperAdmin' ? '' : 'none' }}
          className="btn right"
          onClick={(event) => {
            event.preventDefault();
            $('#title').prop('disabled', (i, v) => (!v));
            $('#title').focus();
            $('#docbody, #docbodyedit, #submitbtn, #editbtn')
              .toggleClass('hide');
          }}
        >edit</button>
        <button
          id="submitbtn"
          className="btn right hide"
          onClick={(event) => {
            event.preventDefault();
            processEdit(editDocument, documentId, history);
            editDocument(documentValue, documentId);
          }}
        >{docStatus}</button>
      </div><br />
    </div>
  );

ReadDocument.propTypes = {
  documentTitle: propTypes.string.isRequired,
  body: propTypes.string.isRequired,
  userId: propTypes.number.isRequired,
  ownerId: propTypes.number.isRequired,
  roleType: propTypes.string.isRequired,
  author: propTypes.string.isRequired,
  modifiedDate: propTypes.string.isRequired,
  editDocument: propTypes.func.isRequired,
  documentId: propTypes.number.isRequired,
  changeTitleValue: propTypes.func.isRequired,
  docStatus: propTypes.string.isRequired,
  history: propTypes.shape({
    push: propTypes.func.isRequired,
  }).isRequired,
};
export default ReadDocument;
