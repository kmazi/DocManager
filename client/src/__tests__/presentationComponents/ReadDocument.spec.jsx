import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ReadDocument from '../../components/presentation/ReadDocument';

describe('The ReadDocument component:', () => {
  const event = {
    preventDefault: () => 'hello',
  };
  const response = { status: 'successful' };
  const props = {
    documentTitle: 'jackson goes to school',
    body: 'I have a friend whose name is jackson',
    author: 'jackson',
    modifiedDate: 'june 20, 2010',
    history: { push: jest.fn() },
    changeTitleValue: jest.fn(),
    editDocument: jest.fn(() => Promise.resolve(response)),
    documentId: 1,
    docStatus: 'successful',
    userId: 5,
    roleType: 'Learning',
    ownerId: 5,
  };
  test('should render the ReadDocument component correctly', () => {
    const enzymeWrapper = shallow(<ReadDocument {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test(`should successfully change the input element for when
   #title input text changes`,
    () => {
      const enzymeWrapper = shallow(<ReadDocument {...props} />);
      enzymeWrapper.find('#readdocument #title').props().onChange();
      expect(props.changeTitleValue.mock.calls.length).toBe(1);
    });

  test('should display form elements to edit the documents',
    () => {
      const enzymeWrapper = shallow(<ReadDocument {...props} />);
      enzymeWrapper.find('#editbtn').props().onClick(event);
      expect(enzymeWrapper.find('#docbody').hasClass('hide')).toBe(false);
      expect(enzymeWrapper.find('#docbodyedit').hasClass('hide')).toBe(true);
      expect(enzymeWrapper.find('#submitbtn').hasClass('hide')).toBe(true);
    });

  test('should successfully edit a document when input value is valid',
    () => {
      const enzymeWrapper = shallow(<ReadDocument {...props} />);
      enzymeWrapper.find('#submitbtn').props().onClick(event);
      expect(props.editDocument.mock.calls.length).toBe(2);
    });

  test('should alert error message when editing document fails', () => {
    response.status = 'unsuccessful';
    const enzymeWrapper = shallow(<ReadDocument {...props} />);
    enzymeWrapper.find('#submitbtn').props().onClick(event);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should display edit button when roleType is admin', () => {
    props.userId = 1;
    props.roleType = 'Admin';
    const enzymeWrapper = shallow(<ReadDocument {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should display edit button when roleType is superadmin', () => {
    props.roleType = 'SuperAdmin';
    const enzymeWrapper = shallow(<ReadDocument {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should not display edit button when roleType is Learning', () => {
    props.roleType = 'Learning';
    const enzymeWrapper = shallow(<ReadDocument {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
