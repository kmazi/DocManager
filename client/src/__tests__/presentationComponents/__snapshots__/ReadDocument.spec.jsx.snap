// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ReadDocument component: should alert error message when editing document fails 1`] = `
<div
  className="container"
  id="readdocument"
>
  <input
    className="center-align"
    disabled={true}
    id="title"
    onChange={[Function]}
    type="text"
    value="jackson goes to school"
  />
  <div
    className=""
    id="docbody"
  >
    I have a friend whose name is jackson
  </div>
  <br />
  <div
    className="hide"
    id="docbodyedit"
  >
    <TinyMCE
      config={
        Object {
          "browser_spellcheck": true,
          "height": 280,
          "plugins": "autolink link image lists print preview",
          "toolbar": "undo redo | bold italic | alignleft aligncenter alignright",
          "width": "100%",
        }
      }
      content="I have a friend whose name is jackson"
      onChange={[Function]}
    />
    <br />
  </div>
  <div>
    <span>
      Author: 
      jackson
    </span>
    <br />
    <span>
      Modified: 
      Invalid Date
    </span>
    <button
      className="btn right"
      onClick={[Function]}
    >
      back
    </button>
    <button
      className="btn right"
      id="editbtn"
      onClick={[Function]}
      style={
        Object {
          "display": "",
        }
      }
    >
      edit
    </button>
    <button
      className="btn right hide"
      id="submitbtn"
      onClick={[Function]}
    >
      successful
    </button>
  </div>
  <br />
</div>
`;

exports[`The ReadDocument component: should display edit button when roleType is admin 1`] = `
<div
  className="container"
  id="readdocument"
>
  <input
    className="center-align"
    disabled={true}
    id="title"
    onChange={[Function]}
    type="text"
    value="jackson goes to school"
  />
  <div
    className=""
    id="docbody"
  >
    I have a friend whose name is jackson
  </div>
  <br />
  <div
    className="hide"
    id="docbodyedit"
  >
    <TinyMCE
      config={
        Object {
          "browser_spellcheck": true,
          "height": 280,
          "plugins": "autolink link image lists print preview",
          "toolbar": "undo redo | bold italic | alignleft aligncenter alignright",
          "width": "100%",
        }
      }
      content="I have a friend whose name is jackson"
      onChange={[Function]}
    />
    <br />
  </div>
  <div>
    <span>
      Author: 
      jackson
    </span>
    <br />
    <span>
      Modified: 
      Invalid Date
    </span>
    <button
      className="btn right"
      onClick={[Function]}
    >
      back
    </button>
    <button
      className="btn right"
      id="editbtn"
      onClick={[Function]}
      style={
        Object {
          "display": "",
        }
      }
    >
      edit
    </button>
    <button
      className="btn right hide"
      id="submitbtn"
      onClick={[Function]}
    >
      successful
    </button>
  </div>
  <br />
</div>
`;

exports[`The ReadDocument component: should display edit button when roleType is superadmin 1`] = `
<div
  className="container"
  id="readdocument"
>
  <input
    className="center-align"
    disabled={true}
    id="title"
    onChange={[Function]}
    type="text"
    value="jackson goes to school"
  />
  <div
    className=""
    id="docbody"
  >
    I have a friend whose name is jackson
  </div>
  <br />
  <div
    className="hide"
    id="docbodyedit"
  >
    <TinyMCE
      config={
        Object {
          "browser_spellcheck": true,
          "height": 280,
          "plugins": "autolink link image lists print preview",
          "toolbar": "undo redo | bold italic | alignleft aligncenter alignright",
          "width": "100%",
        }
      }
      content="I have a friend whose name is jackson"
      onChange={[Function]}
    />
    <br />
  </div>
  <div>
    <span>
      Author: 
      jackson
    </span>
    <br />
    <span>
      Modified: 
      Invalid Date
    </span>
    <button
      className="btn right"
      onClick={[Function]}
    >
      back
    </button>
    <button
      className="btn right"
      id="editbtn"
      onClick={[Function]}
      style={
        Object {
          "display": "",
        }
      }
    >
      edit
    </button>
    <button
      className="btn right hide"
      id="submitbtn"
      onClick={[Function]}
    >
      successful
    </button>
  </div>
  <br />
</div>
`;

exports[`The ReadDocument component: should not display edit button when roleType is Learning 1`] = `
<div
  className="container"
  id="readdocument"
>
  <input
    className="center-align"
    disabled={true}
    id="title"
    onChange={[Function]}
    type="text"
    value="jackson goes to school"
  />
  <div
    className=""
    id="docbody"
  >
    I have a friend whose name is jackson
  </div>
  <br />
  <div
    className="hide"
    id="docbodyedit"
  >
    <TinyMCE
      config={
        Object {
          "browser_spellcheck": true,
          "height": 280,
          "plugins": "autolink link image lists print preview",
          "toolbar": "undo redo | bold italic | alignleft aligncenter alignright",
          "width": "100%",
        }
      }
      content="I have a friend whose name is jackson"
      onChange={[Function]}
    />
    <br />
  </div>
  <div>
    <span>
      Author: 
      jackson
    </span>
    <br />
    <span>
      Modified: 
      Invalid Date
    </span>
    <button
      className="btn right"
      onClick={[Function]}
    >
      back
    </button>
    <button
      className="btn right"
      id="editbtn"
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
    >
      edit
    </button>
    <button
      className="btn right hide"
      id="submitbtn"
      onClick={[Function]}
    >
      successful
    </button>
  </div>
  <br />
</div>
`;

exports[`The ReadDocument component: should render the ReadDocument component correctly 1`] = `
<div
  className="container"
  id="readdocument"
>
  <input
    className="center-align"
    disabled={true}
    id="title"
    onChange={[Function]}
    type="text"
    value="jackson goes to school"
  />
  <div
    className=""
    id="docbody"
  >
    I have a friend whose name is jackson
  </div>
  <br />
  <div
    className="hide"
    id="docbodyedit"
  >
    <TinyMCE
      config={
        Object {
          "browser_spellcheck": true,
          "height": 280,
          "plugins": "autolink link image lists print preview",
          "toolbar": "undo redo | bold italic | alignleft aligncenter alignright",
          "width": "100%",
        }
      }
      content="I have a friend whose name is jackson"
      onChange={[Function]}
    />
    <br />
  </div>
  <div>
    <span>
      Author: 
      jackson
    </span>
    <br />
    <span>
      Modified: 
      Invalid Date
    </span>
    <button
      className="btn right"
      onClick={[Function]}
    >
      back
    </button>
    <button
      className="btn right"
      id="editbtn"
      onClick={[Function]}
      style={
        Object {
          "display": "",
        }
      }
    >
      edit
    </button>
    <button
      className="btn right hide"
      id="submitbtn"
      onClick={[Function]}
    >
      successful
    </button>
  </div>
  <br />
</div>
`;
